### PR TITLE
KEP-388, KEP-389: Audit module liveness checks

### PR DESCRIPTION
**Pull request by isabelsavannah**
Please pay particular attention to the audit_test class. My goal there is to avoid duplication by allowing derived tests to declare different expectations on mock timers without duplicating the code that creates and injects those timers. It works, but I'm not completely sure on the safety/best-practices-ness of it.

### CODE REVIEW

File: audit/audit.cpp [hunk 1]
```diff
@@ -46,9 +48,74 @@ audit::start()
...
+    this->leader_alive_timer->expires_from_now(this->leader_timeout);
+    this->leader_alive_timer->async_wait(std::bind(&audit::handle_leader_alive_timeout, shared_from_this(), std::placeholders::_1));
+}
+
+void
+audit::handle_leader_alive_timeout(const boost::system::error_code& ec)
+{
+    if(ec)
+    {
+        LOG(debug) << "Leader alive timeout canceled " << ec;
```
Comment (ebruck) [inline audit/audit.cpp]: use: ec.message() instead of the error code.

File: audit/audit.hpp [hunk 2]
```diff
@@ -38,13 +40,37 @@ namespace bzn
...
+
         std::map<uint64_t, bzn::uuid_t> recorded_leaders;
         std::map<uint64_t, std::string> recorded_commits;
 
         std::once_flag start_once;
+        std::mutex audit_lock;
+        std::unique_ptr<bzn::asio::steady_timer_base> leader_alive_timer;
+        std::unique_ptr<bzn::asio::steady_timer_base> leader_progress_timer;
+
+        std::chrono::milliseconds leader_timeout{std::chrono::milliseconds(30000)};
```
Comment (ebruck) [inline audit/audit.hpp]: Is this something we may want to configure?

File: audit/audit.hpp [hunk 3]
```diff
@@ -38,13 +40,37 @@ namespace bzn
...
+
+        bzn::uuid_t last_leader;
+        uint64_t last_leader_commit_index;
```
Comment (ebruck) [inline audit/audit.hpp]: Set this to zero?

File: audit/audit.hpp (hunk 2 above)
Comment (isabelsavannah) [reply audit/audit.hpp]: Yeah, but I was not including that in this commit to make it as simple as possible for the sake of Lovelace

**Pull request merged**